File: leetCode/1021removeOuterParentheses/solution.js
/**
 * @param {string} S
 * @return {string}
 */
var removeOuterParentheses = function(S) {
    var ans ="",stack = [];
    for(var char of S){
        if(char == '('){
            if(stack.length>0){
                ans+=char;
            }
            stack.push(char);
        }else if(char == ')'){
            stack.pop();
            if(stack.length>0){
                ans+=char;
            }
        }else{
            ans+=char;
        }
    } 
    console.log(ans);
    return ans;
};
// var removeOuterParentheses = function(S) {
//     class Stack{
//         constructor(){
//             this.data = [];
//         }

//         pop(){
//             if(!this.isEmpty()){
//                 return this.data.pop();
//             } 
//         }

//         push(val){
//             this.data.push(val);
//         }

//         isEmpty(){
//             return this.data<=0;
//         }
//     }

//     var ans ="";
//     var stack = new Stack();
//     for(var char of S){
//         if(char == '('){
//             if(!stack.isEmpty()){
//                 ans+=char;
//             }
//             stack.push(char);
//         }else if(char == ')'){
//             stack.pop();
//             if(!stack.isEmpty()){
//                 ans+=char;
//             }
//         }else{
//             ans+=char;
//         }
//     } 
//     return ans;
// };

removeOuterParentheses("(()())(())");

File: leetCode/1021removeOuterParentheses/solution.java
class Solution {
    public String removeOuterParentheses(String S) {
        int flag = 0;
        StringBuilder ans = new StringBuilder();
        for(int i=0;i< S.length();i++){
            char temp = S.charAt(i);
            if(temp == '(' && flag++ > 0){
                ans.append(temp);
            }else if(temp == ')' &&  --flag>0){
                ans.append(temp);
            }
        } 
        return ans.toString();
    }
}

// class Solution {
//     public String removeOuterParentheses(String S) {
//         int flag = 0;
//         StringBuilder ans = new StringBuilder();
//         for(int i=0;i< S.length();i++){
//             char temp = S.charAt(i);
//             if(temp == '('){
//                 if(flag>0){
//                     ans.append(temp);
//                 }
//                 flag++;
//             }else if(temp == ')'){
//                 flag--;
//                 if(flag>0){
//                     ans.append(temp);
//                 }
//             }
//         } 
//         return ans.toString();
//     }
// }

// import java.util.Stack;

// class Solution {
//     public String removeOuterParentheses(String S) {
//         Stack<Character> stack = new Stack<>();
//         StringBuilder ans = new StringBuilder();
//         for(int i=0;i< S.length();i++){
//             char temp = S.charAt(i);
//             if(temp == '('){
//                 if(!stack.empty()){
//                     ans.append(temp);
//                 }
//                 stack.push(temp);
//             }else if(temp == ')'){
//                 stack.pop();
//                 if(!stack.empty()){
//                     ans.append(temp);
//                 }
//             }else{
//                 ans.append(temp);
//             }
//         } 
//         return ans.toString();
//     }
// }